### PR TITLE
WIP: Refactor Snap to support a Hook pattern

### PR DESCRIPTION
**Pull request by Zombiefruit**
## Describe this Pull Request
The goal was to refactor snap to allow hooking into different stages of the snap process (project, component, and simulation) in order to make snap more generic, and allow different projects to implement their own logic file -> html file flow.

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] None of the above

### Versioning required due to this change
- [ ] Patch
- [ ] Minor
- [x] Major

### Does this change the API or main functionality
- [x] Yes and I updated README.md
- [ ] No

# URG MAKE SNAP CONFIG
Snap use **PLUGINS**

plugins **GOOD**

plugins have internal state since they are just classes. Mean that plugins can do whatever they want 

## CONFIG FILE

``` js
const {myPlugin} = require('./my-plugin.js');

module.exports = {
  applitoolsApiKeyName: ‘URGS_KEY_NAME’,
  plugins: [myPlugin]
}
```

plugin interface look like this:

```ts
interface ISnapPlugin {
  projectHook?: (snapInfo: ISnapInfo) => Promise<void>;
  componentHook?: (snapInfo: ICompSnapInfo) => Promise<IFileInfo | void>;
  simHook?: (snapInfo: ISimSnapInfo) => Promise<IFileInfo | void>;
  afterHook?: (snapInfo: ISnapInfo, files: IFileInfo[]) => Promise<void>;
}

```

Urg plugin look like this:

```ts
class MyPlugin implements ISnapPlugin {
  constructor(private urgPrivateVariable: any) {}
    
  public projectHook = async (snapInfo: ISnapInfo): Promise<void> => {...}
    
  public simHook = async (simInfo: ISimSnapInfo): Promise<void> => {...}
    
  public afterHook = async (snapInfo: ISnapInfo): Promise<void> => {...}
}

```

Urg wrote plugin himself. Urg like typescript. Urg no like not typescript.

Urg say that project write their own plugins. Easy. Mean that Urg no maintain plugins. Urg like this.

Urg change metadata. Make function called `addExportInfo`.

Now Component Metadata look like this:

```ts
interface IComponentMetadata {
  // ...
  exportInfo: Partial<IExportInfo> | any;
}

export interface IExportInfo {
  path: string;
  exportName: string;
  baseStylePath: string;
  cssPath: string;
  compPath: string;
}
```

Urg make it so project can add whatever they want to `exportInfo`, so make easy for customize in future. Far future. Maybe even a time when URG and his tribe are gone away. Urg no like to think about nature of time. So fleeting.

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -190,11 +190,487 @@ autotools-snap --files ./components/**/*.meta.ts
...
+```
+All fields are optional. However, without hooks, Snap won't do anything.
+
+#### Building
+
+Each project may have a different or unique build step. Components may be Stylable-based, or use Sass, or css-modules, or have special imports, or be written in Typescript... the list goes on. In order to keep Snap generic, and usable in many different situations, we've implemented what we're calling *hooks*.
+
+Hooks allow a project to build itself, and then give Snap the information it needs in order to provide screenshot regression tests.
+
+This is our contract, in its simplest terms: Snap will call all of the hooks in the config in a specific order, passing information about the project, and after all the hooks have been called, Snap expects to receive a list containing the locations of HTML files which need to be sent to Applitools. Luckily, we already provide some hooks that cover common-use cases, so as a project, you may not need to write your own.
```
Comment (JoMarton) [inline README.md]: We may be missing basic background about how snap works - from simulation to HTML/CSS to screenshot test

File: README.md [hunk 2]
```diff
@@ -190,11 +190,487 @@ autotools-snap --files ./components/**/*.meta.ts
...
+  Registry: IRegistry;
+  projectPath: string;
+  tempDirPath: string;
+  collectFiles: (files: IFileInfo[]) => void;
+}
+```
+
+`ISnapInfo` contains: `Registry`, which will contain all the project metadata at the point it is passed to each hook. `projectPath` describes the base path of the project in which snap is being run. `tempDirPath` describes the path to a temporary directory where snapshot files (HTML files) should be written to. `collectFiles` is a method used by hooks to append files to the list which will be passed to Applitools.
+
+Project hooks must return a Promise, but that Promise is of type `any`. Whatever is returned from a project hook will be stored in an object called `projectHookData`, which is then passed to each hook. This is useful in certain cases, for example, if you wanted to create a temporary directory in your project hook, store generated files there, and then destroy the directory in the after hook.
```
Comment (Zombiefruit) [inline README.md]: This isn't the case right now (`projectHookData` is only passed to After hooks). Maybe this should be changed.

File: README.md [hunk 3]
```diff
@@ -190,11 +190,487 @@ autotools-snap --files ./components/**/*.meta.ts
...
+Configuration is as simple as adding a `autotools.config.js` file in your `.autotools` folder.
+If you wish to use a configuration file named something different, or located somewhere else, you can specify a specific file using the `--config` flag, like so: 
+
+`yarn snap --config ./config/myConfig.js`
+
+The format of the `autotools.config.js` file is as follows:
+
+``` js
+module.exports = {
+    applitoolsApiKeyName: ‘WHATEVER_YOU_NAMED_YOUR_ENV_VARIABLE’,
```
Comment (JoMarton) [inline README.md]: line 177 says: "This tool requires that the `process.env.EYES_API_KEY` value is set to your private API key."

File: README.md [hunk 4]
```diff
@@ -190,11 +190,488 @@ autotools-snap --files ./components/**/*.meta.ts
...
+
+- **project**: called once, at the beginning. This is for setup (creating temporary directories, building the project if it only needs to be built once, etc.) 
+- **component**: called once for every component in the project which has a metadata file.
+- **sim**: called once for each simulation of every component.
+- **after**: called once after all other hooks have been called. This is where you'd do any necessary cleanup (deleting temporary directories, files, etc.)
+
+The interface for `IFileInfo` is as follows:
+
+```ts
+interface IFileInfo {
```
Comment (JoMarton) [inline README.md]: maybe add inline comments describing each attribute? the example isn't enough i feel

File: README.md [hunk 5]
```diff
@@ -190,11 +190,488 @@ autotools-snap --files ./components/**/*.meta.ts
...
+{
+    basename: 'blue.ts',
+    filepath: 'users/JimBob/Documents/colors/blue.ts',
+    data: 'import {red} from \'./red\';\nlet violet...',
+    cssPath: 'users/JimBob/Documents/colors/colors.css'
+}
+
+```
+
+The interface for the hook arguments (`ISnapInfo`, `ICompSnapInfo`, etc.) is described in the following section.
```
Comment (JoMarton) [inline README.md]: Not sure this line is necessary

File: README.md [hunk 6]
```diff
@@ -190,11 +190,488 @@ autotools-snap --files ./components/**/*.meta.ts
...
+- `stylable-snap-plugin`: for building and snapping stylable projects.
+- `wix-style-react-snap-plugin`: for wix/wix-style-react.
+
+###### Project Hook
+
+The project hooks are run at the beginning of Snap execution, and are passed an object of type `ISnapInfo`.
+
+``` ts
+interface ISnapInfo {
+  Registry: IRegistry;
```
Comment (JoMarton) [inline README.md]: capital R on purpose?

File: README.md [hunk 7]
```diff
@@ -190,11 +190,488 @@ autotools-snap --files ./components/**/*.meta.ts
...
+Project hooks must return a Promise, but that Promise is of type `any`. Whatever is returned from a project hook will be stored in an object called `projectHookData`, which is then passed to each hook. This is useful in certain cases, for example, if you wanted to create a temporary directory in your project hook, store generated files there, and then destroy the directory in the after hook.
+
+###### Component Hook
+
+Component hooks are run once for every component in the project which has an associated metadata file. Component hooks are passed an `ICompSnapInfo` object:
+
+```ts
+interface ICompSnapInfo extends ISnapInfo {
+  componentMetadata: IComponentMetadata<any, any>;
+  projectHookData: any;
```
Comment (JoMarton) [inline README.md]: given your comment that `projectHookData` currently only gets to After hooks - should this stay?

File: packages/mock-repo/.autotools/mock-repo-plugin.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+const asyncExec = util.promisify(exec);
+
+export class MockRepoPlugin implements ISnapPlugin {
+  /**
+   * This is the key for a custom field in the metadata
+   * where we'll store any information we need for this plugin
+   */
+  public static metadataKey = 'MockRepoPluginInfo';
+  public buildDirectory = 'dist';
```
Comment (alisey) [inline packages/mock-repo/.autotools/mock-repo-plugin.ts]: Make it private?

File: packages/mock-repo/.autotools/mock-repo-plugin.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+  /**
+   * Project hooks are run at the beginning of Snap execution. This is where
+   * we'll do any necessary setup. In this case, we want to build the project.
+   */
+  public projectHook  = async (): Promise<void> => {
+    const {stdout, stderr} = await asyncExec('yarn build');
+    if (stdout) { consoleLog('stdout:', stdout); }
+    if (stderr) { consoleLog('stderr:', stderr); }
```
Comment (alisey) [inline packages/mock-repo/.autotools/mock-repo-plugin.ts]: Should throw.

File: packages/mock-repo/.autotools/mock-repo-plugin.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        for (const resource of componentMetadata.resources) {
+          snapshotResources.push({
+            url: resource.url,
+            mimeType: resource.mimeType,
+            data: fs.readFileSync(path.join(simInfo.projectPath, resource.absolutePath))
+          });
+
+          if (resource.mimeType === 'text/css') {
+            // Then we want to include a link in the head
+            links.push(`<link rel="stylesheet" type="text/css" href="${resource.url}">`);
```
Comment (alisey) [inline packages/mock-repo/.autotools/mock-repo-plugin.ts]: Need to use `encodeURI()`

File: packages/mock-repo/.autotools/mock-repo-plugin.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+      /**
+       * Components may have resources, which we need to read from disk so that our snaphsot files can reference
+       * the resources in tests.
+       */
+      if (componentMetadata.resources) {
+        for (const resource of componentMetadata.resources) {
+          snapshotResources.push({
+            url: resource.url,
+            mimeType: resource.mimeType,
+            data: fs.readFileSync(path.join(simInfo.projectPath, resource.absolutePath))
```
Comment (alisey) [inline packages/mock-repo/.autotools/mock-repo-plugin.ts]: Wait, if it's an absolute path, why do we join it with something else?

File: packages/snap/src/snap.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        if (plugin.projectHook) {
+          await plugin.projectHook(snapInfo);
+        }
+
+        await runComponentAndSimHooks(snapInfo, plugin);
+      }
+
+      if (snapshotFiles.length === 0) {
+        // Maybe a better message
+        consoleError('No files were found.');
```
Comment (alisey) [inline packages/snap/src/snap.ts]: Maybe a better message?

File: packages/snap/src/utils/eyes-key-exists.ts [hunk 13]
```diff
@@ -5,7 +5,7 @@ export function eyesKeyExists(): boolean {
     return false;
   }
 
-  if (!process.env.APPLITOOLS_API_KEY) {
+  if (!process.env.APPLITOOLS_API_KEY && !process.env.applitoolsApiKeyName) {
```
Comment (alisey) [inline packages/snap/src/utils/eyes-key-exists.ts]: Is this outdated?

File: packages/snap/src/utils/filename-utils.ts [hunk 14]
```diff
@@ -1,14 +1,13 @@
 export interface IFileParts {
   compName: string;
-  simIndex: number;
   simName: string;
   styleIndex?: number;
   styleName?: string;
 }
 
-export function generateSnapshotFilename(componentName: string, simName: string, simIndex: number, styleName?: string): string {
+export function generateSnapshotFilename(componentName: string, simName: string, styleName?: string): string {
```
Comment (alisey) [inline packages/snap/src/utils/filename-utils.ts]: If this is only related to stylable move it to the stylable plugin folder.

File: packages/snap/src/utils/generate-snapshots.ts [hunk 15]
```diff
@@ -4,52 +4,31 @@ import webpack from 'webpack';
...
-    }
-  }
+export interface IFileInfo {
+  basename: string;
+  filepath: string;
 }
 
-export function render(fileName: string, Registry: IRegistry, compiledFile: any, sourceFile: any) {
-  const compMetadata = findComponentByName(compiledFile.default.name, Registry);
+export function render(fileName: string, snapInfo: ISimSnapInfo, returnHtml: (html: string) => void, compiledFile: any, _sourceFile: any) {
```
Comment (alisey) [inline packages/snap/src/utils/generate-snapshots.ts]: If this is only related to stylable move it to the stylable folder. Why does `_sourceFile` have an underline?

File: README.md [hunk 16]
```diff
@@ -50,6 +50,15 @@ import theme1 from './theme1.st.css';
 // and returns the newly created metadata
 const myComponentMetadata = Registry.getComponentMetadata(MyComp);
 
+// You can add resources (currently only used by Snap)
+myComponentMetadata.resources = [
```
[the remark this answers is not in the document]
Comment (Zombiefruit) [reply README.md]: Ah, I renamed the type but not the field

File: packages/registry/src/registry/component-metadata.tsx [hunk 17]
```diff
@@ -9,9 +9,9 @@ export class ComponentMetadata<Props, State> implements IComponentMetadata<Props
   public nonA11yCompliant: boolean = false;
   public nonEventListenerTestCompliant: boolean = false;
   public nonHydrationTestCompliant: boolean = false;
-  public path: string = '';
-  public exportName: string = '';
-  public baseStylePath: string = '';
+  public exportInfo: Partial<IExportInfo> = {};
+  public resources: IStaticResource[] = [];
+  public customFields: any = {};
```
[the remark this answers is not in the document]
Comment (JoMarton) [reply packages/registry/src/registry/component-metadata.tsx]: Seems like it should also be private if adding is done through a method

File: packages/registry/src/tsconfig.json [hunk 18]
```diff
@@ -1,3 +1,3 @@
 {
-  "extends": "../../../tsconfig.base.json"
+  "extends": "../../../tsconfig.base.json",
```
Comment (JoMarton) [inline packages/registry/src/tsconfig.json]: comma

File: packages/snap/src/utils/eyes-key-exists.ts (hunk 13 above)
Comment (Zombiefruit) [reply packages/snap/src/utils/eyes-key-exists.ts]: Whoops, yep

File: packages/snap/src/utils/generate-snapshots.ts (hunk 15 above)
Comment (Zombiefruit) [reply packages/snap/src/utils/generate-snapshots.ts]: It's unused but we need a param after it. The _ makes the linter happy

File: packages/mock-repo/.autotools/mock-repo-plugin.ts [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+      }
+
+      const html = createHtml(compRenderedToString, links, simulationName);
+      const file: ISnapshot = {html, testName, staticResources: snapshotResources};
+
+      /**
+       * Snap provides a method which we can call with one or more files. Snap stores these files,
+       * and only sends them to Applitools after every component and simulation hook has run
+       */
+      simInfo.collectFiles([file]);
```
Comment (alisey) [inline packages/mock-repo/.autotools/mock-repo-plugin.ts]: How about `simInfo.collectSnapshot(snapshot)`?

File: packages/mock-repo/src/css-comp/css-comp.meta.ts [hunk 20]
```diff
@@ -1,13 +1,30 @@
...
 const metadata = Registry.getComponentMetadata(CssComp);
+
+metadata.exportInfo = {
+    path: 'src/css-comp/css-comp.tsx',
+    exportName: 'CssComp'
+};
+
+metadata.addCustomField(MockRepoPlugin.metadataKey, {
+    compPath: 'cjs/css-comp/css-comp',
+    cssPath: 'src/css-comp/css-comp.css',
```
Comment (alisey) [inline packages/mock-repo/src/css-comp/css-comp.meta.ts]: The plugin is not using this property. Let's remove it.

File: packages/mock-repo/.autotools/mock-repo-plugin.ts [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+          snapshotResources.push({
+            url: resource.url,
+            mimeType: resource.mimeType,
+            data: fs.readFileSync(path.join(simInfo.projectPath, resource.path))
+          });
+        }
+      }
+
+      const html = createHtml(compRenderedToString, links, simulationName);
+      const file: ISnapshot = {html, testName, staticResources: snapshotResources};
```
[the remark this answers is not in the document]
Comment (Zombiefruit) [reply packages/mock-repo/.autotools/mock-repo-plugin.ts]: doh, the interface name was a good hint..

File: packages/mock-repo/src/image-comp/image-comp.tsx [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+    src: string;
+}
+
+export default class ImageComp extends React.Component<IProps> {
+  public static displayName = 'ImageComp';
+
+  public render() {
+    return (
+    <div className="root">
+      <img alt="image of a lonely chair in a lonely room" src={this.props.src} className="image" />
```
Comment (alisey) [inline packages/mock-repo/src/image-comp/image-comp.tsx]:
```suggestion
      <img alt="image of a lonely chair in a lonely room, alone" src={this.props.src} className="image" />
```

File: packages/registry/src/registry/component-metadata.tsx [hunk 23]
```diff
@@ -9,9 +9,9 @@ export class ComponentMetadata<Props, State> implements IComponentMetadata<Props
   public nonA11yCompliant: boolean = false;
   public nonEventListenerTestCompliant: boolean = false;
   public nonHydrationTestCompliant: boolean = false;
-  public path: string = '';
-  public exportName: string = '';
-  public baseStylePath: string = '';
+  public exportInfo: IExportInfo = {path: '', exportName: ''};
```
Comment (alisey) [inline packages/registry/src/registry/component-metadata.tsx]: I'd prefer `IExportInfo | null`. Seems like `if (metadata.exportInfo) { /* use it */ }` is a bug waiting to happen.

Comment (Zombiefruit) [reply packages/registry/src/registry/component-metadata.tsx]: Great, will change

**Pull request merged**